File: Data Structure/Array/NCPC Preliminary/H.cpp
#include<bits/stdc++.h>
using namespace std;

int dayNumber(int day, int month, int year) {
    static int t[] = { 0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4 };
    year -= month < 3;
    return (year + year/4 - year/100 + year/400 + t[month-1] + day) % 7;
}

string getMonthName(int monthNumber) {
    string months[] = {"January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"};
    return (months[monthNumber]);
}

int numberOfDays(int monthNumber, int year) {
    switch(monthNumber) {
        case 1:
            return (year % 400 == 0 || (year % 4 == 0 && year %100 != 0)) ? 29 : 28;
            break;
        case 0:
        case 2:
        case 4:
        case 6:
        case 7:
        case 9:
        case 11: return(31);
            break;
        
        case 3:
        case 5:
        case 8:
        case 10: return(30);
            break;
    }
}

void printCalendar(int month, int year) {
    cout << "|---------------------------|" << endl;
    cout << "|Sun|Mon|Tue|Wed|Thu|Fri|Sat|" << endl;
    cout << "|---------------------------|" << endl;

    int days = numberOfDays(month, year);
    int current = dayNumber(1, month, year);
    int k = 0;

    for (int i = 0; k < days; i++) {
        cout << "|";
        for (int j = 0; j < 7; j++) {
            if ((i == 0 && j < current) || k >= days) {
                cout << " - |";
            } else {
                printf("%3d|", k + 1);
                k++;
            }
        }
        cout << endl << "|---------------------------|" << endl;
    }
}

int main() {
    int t; 
    cin >> t;

    while (t--) {
        int day, month, year;
        char dummy;
        
        cin >> day >> dummy >> month >> dummy >> year;

        printCalendar(month, year);
        cout<<endl;
    }

    return 0;
}

File: Data Structure/Array/NCPC Preliminary/h.py
from calendar import monthrange, day_name

def print_calendar(day, month, year, day_name):
    # Calculate the number of days in the month
    num_days = monthrange(year, month)[1]

    # Create a matrix to represent the calendar
    calendar_matrix = [[' - ' for _ in range(7)] for _ in range(5)]

    # Find the starting day of the week for the given date
    start_day = day_name.index(day_name)

    # Populate the matrix with the corresponding days of the month
    current_day = 1
    for i in range(5):
        for j in range(7):
            if i == 0 and j < start_day:
                continue
            if current_day <= num_days:
                calendar_matrix[i][j] = f"{current_day:2d}"
                current_day += 1

    # Print the calendar in tabular format
    print("|---------------------------|")
    print("|Sun|Mon|Tue|Wed|Thu|Fri|Sat|")
    print("|---------------------------|")
    for row in calendar_matrix:
        print("|" + "|".join(row) + "|")
        print("|---------------------------|")

# Read the number of test cases
T = int(input().strip())

for _ in range(T):
    # Read the input for each test case
    date_input = input().strip().split()
    day, month, year, day_name = map(int, date_input[0].split('/')) + [date_input[1]]

    # Print the calendar for the given date
    print_calendar(day, month, year, day_name)
    print()  # Print a blank line after each test case
